Add python script to update benchmarks from Azure output

NeilZaim opened the pull request:
As discussed in last week's dev meeting, I upload here the python script that I've been using to update the benchmarks from Azure outputs (that way I don't have to compile&run the code locally). The script is not very well written (lots of magic numbers and the parsing is certainly not robust to changes in the Azure output or in the way we format the benchmark json files), but it's a small and easy to debug script anyways. Let me know if you think that it can be a useful addition to the repo. If so, I'll add the description below to the developers documentation (and the script itself can be improved as well if needed).

**How to use the script:**

1. On the Github page of the Pull Request, find (one of) the test failing due to benchmarks that need to be updated and click on "Details".
![Screenshot1](https://user-images.githubusercontent.com/49716072/135133589-1fd8a626-ff93-4b9e-983f-acee028e0e4e.png)
2. Click on "View more details on Azure pipelines".
![Screenshot2](https://user-images.githubusercontent.com/49716072/135133596-8f73afa2-969e-49a4-b4a6-184a4f478a44.png)
3. Click on "Build & test".
![Screenshot3](https://user-images.githubusercontent.com/49716072/135133607-87324124-6145-4589-9a92-dcc8ea9432e4.png)
4. Click on "View raw log".
![Screenshot4](https://user-images.githubusercontent.com/49716072/135133617-764b6daf-a8e4-4a50-afae-d4b3a7568b2f.png)
5. This should lead to a page that looks like the image below. Save it as a text file on your local computer.
![Screenshot5](https://user-images.githubusercontent.com/49716072/135133624-310df207-5f87-4260-9917-26d5af665d60.png)
6. On your local computer, go to the WarpX folder and cd to the `Regression/Checksum/` folder.
7. Run the command `python update_benchmarks_from_azure_output.py  /path/to/azure_output.txt`. The benchmarks included in that Azure output should now be updated.
8. Repeat this for every Azure pipeline (e.g. `cartesian2d`, `cartesian3d`, `qed`) that contains benchmarks that need to be updated.
9. Commit the changes.

ax3l commented:
That is brilliant.

If we add a tiny function around it, we can even make this a script that takes a pull request number or link and follows the links automatically... :)

dpgrote commented:
This is a great idea - thanks! I had the same thought as Axel - can this grab the log files? Does `GitHub` have an API for this?

Other comments - it is standard in Python now to do any file access within a `with` clause. For example:

`with open(azure_output_filename, "r") as f:
    for line in f:
        ...
`

Also, instead of hand parsing the json, can you use the `json` Python module? This is what is used in `Regression/Checksum/benchmark.py`. Your script can load the json file, modify the appropriate parameters, and then do a `json.dump`.

EZoni commented on Regression/Checksum/update_benchmarks_from_azure_output.py:
```diff
@@ -0,0 +1,48 @@
+
+azure_output_filename = sys.argv[1]
+
+first_char = 29 # first non prefix character
+test_name_str = "working on test:"
+test_name_first_char = 50 # start of test name
+test_name_last_char = -5 # end of test name
```
I am a bit puzzled by some of these numbers. Looking at the raw logs, there seems to be usually an initial alphanumeric string (containing info like date, time, etc.) such as `2000-01-01T00:00:00.1234567A` (28 characters plus a trailing white space, so 29 characters in total). Then I seem to see a few weird characters such as a symbol plus `[1m` (another 4 characters), before the string `working on test:` starts. However, these additional 4 characters seem not to be included in your counting, when you set `first_char = 29` and `test_name_first_char = 50`. I see these additional characters both when I inspect that raw log on my browser and when I save to a text file locally.

NeilZaim replied:
I think there's two answers to that one:
- The short one: actually I didn't really try to understand those numbers and just found empirically what worked in my python script. :smile:
- The long one: these weird characters are here for formatting (in this case, making the text bold when viewing the output). I'm not exactly sure where your count is off by 4 characters compared to mine, but maybe it's because `'\x1b'` is read as a single character in a python string? In any case, I'm showing below the output of what I get when reading one of these lines with python, so that you can see why I chose 50 here.

![image](https://user-images.githubusercontent.com/49716072/135836164-ab635584-779d-4b57-8202-8f3938982c95.png)

EZoni commented on Regression/Checksum/update_benchmarks_from_azure_output.py:
```diff
@@ -0,0 +1,48 @@
+    if (line.find(test_name_str)!= -1):
+        current_test = line[test_name_first_char:test_name_last_char]
+    line_noprefix = line[first_char:]
+    if (line_noprefix.startswith(plotfile_str)):
```
Here we could add a comment like the following:
```suggestion
    # Here we parse lines that read, for example,
    # "Plotfile : [lev=0,jx] 1.012345678901234e+16"
    # and we set key1 = "lev=0", key2 = "jx", value = 1.012345678901234e+16
    if (line_noprefix.startswith(plotfile_str)):
```

EZoni reviewed (commented):
Thank you for this PR, Neil! In general, if possible I would add a few more inline comments in the code, including an initial short docstring explaining how to run the script (I guess similar to what you plan to add in the online documentation). I suggested examples of such inline comments below (but I left out the `json` part in case you end up modifying it according to what Dave suggested). I also pointed out a question regarding some of the initial "magic" numbers that you set at the beginning of the script.

NeilZaim commented:
Hi all and thanks a lot for your useful comments. I've modified the script following your suggestions:
- I'm now using the `with` constructs to open the files.
- I'm now using `json` module to update the json files. That actually makes things much simpler!
- I've added some inline comments explaining the script.

Also, I agree that it would be great to be able to automatically download the raw output from Azure. @ax3l what solution did you have in mind to do that? I don't have much experience with these sort of things. Using some API to directly download Azure devops log files (couldn't find one but that doesn't mind it doesn't exist), or using a module like `scrapy` or similar to retrieve the data from the Azure website?

dpgrote commented on Regression/Checksum/update_benchmarks_from_azure_output.py:
```diff
@@ -0,0 +1,38 @@
+        # Here we parse lines that read, for example,
+        # "working on test: bilinear_filter"
+        # and we set current_test = "bilinear_filter"
+        if (line.find(test_name_str)!= -1):
```
This can be done using `re` instead of relying on the empirical numbers. Something like
`m = re.search('working on test: (?P<testname>\w*)', line)`. Then you can do `if m: current_test = m.group('testname')`.

dpgrote commented on Regression/Checksum/update_benchmarks_from_azure_output.py:
```diff
@@ -0,0 +1,38 @@
+        # Here we parse lines that read, for example,
+        # "Plotfile : [lev=0,jx] 1.012345678901234e+16"
+        # and we set key1 = "lev=0", key2 = "jx", value = 1.012345678901234e+16
+        if (line_noprefix.startswith(plotfile_str)):
```
The searches here can be combined in one search, starting with "Plotfile : ". This would obviate the need for stripping out the starting part of the line - the search could be done on the whole line. No empirical numbers would be needed.

dpgrote commented on Regression/Checksum/update_benchmarks_from_azure_output.py:
```diff
@@ -0,0 +1,38 @@
+            value = float(re.search('\](.*)', line_noprefix).group(1)[1:])
+            # Update json file with the new value
+            json_filepath = benchmark_path+current_test+benchmark_suffix
+            with open(json_filepath, "r") as f_json:
```
It's a small thing, but perhaps load the json file before looping over lines and dumping it afterwards so the file is read in and written out once.

dpgrote commented:
This looks much better, though I have more comments that will help clean it up some more and make it more robust.

NeilZaim commented:
Thanks again @dpgrote, I've tried to do as you suggested, let me know if that looks good to you.

RemiLehe commented:
@NeilZaim Thanks for this script, this is very useful.
When using it, I noticed that it does parse the test `momentum-conserving-gather` properly (maybe because of the `-` character?). Do you think that there is an easy fix?

NeilZaim commented:
Thanks for finding this bug @RemiLehe. It was indeed because of the `-` character.

I've pushed a quick fix, which maybe is a bit too manual but seems to work. (maybe it would be more robust to parse everything that is until the `\x1b[0m` character that marks the end of the bold section)

roelof-groenewald commented:
I tested this script recently in updating several benchmarks for #3711.
It still works well, so I think it is worth merging this and having the script available without needing to look back for this PR.

NeilZaim commented:
cc. @ax3l if you have time for a review

roelof-groenewald commented on Docs/source/developers/checksum.rst:
```diff
@@ -61,3 +64,52 @@ Since this will automatically change the JSON file stored on the repo, make a se
+Reset a benchmark from the Azure pipeline output on Github
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Alternately, the benchmarks can be reset using the output of the Azure continuous intergration (CI) tests on Github. The output can be accessed by following the steps below:
```
```suggestion
Alternatively, the benchmarks can be reset using the output of the Azure continuous intergration (CI) tests on Github. The output can be accessed by following the steps below:
```

roelof-groenewald commented on Tools/DevUtils/update_benchmarks_from_azure_output.py:
```diff
@@ -0,0 +1,65 @@
+
+In the Azure output, we read the lines contained between
+"New file for Test_Name:"
+and the next occurence of
```
```suggestion
and the next occurrence of
```

roelof-groenewald reviewed (approved):
Thanks again for updating this! It will be very helpful next time there is a change that requires updating many benchmark values.
